src/arm/windows/init-by-logical-sys-info.c

```diff
@@ -0,0 +1,879 @@
+	}
+
+	for (uint32_t i = 0; i <= info->Cache.GroupCount; i++) {
+	/* Zero GroupCount is valid, GroupMask still can store bits set. */
+		const uint32_t group_id = info->Cache.GroupMasks[i].Group;
+		/* Bitmask representing processors in this group belonging to this package */
+		KAFFINITY group_processors_mask = info->Cache.GroupMasks[i].Mask;
+		while (group_processors_mask != 0) {
+			const uint32_t processor_id_in_group =
+				low_index_from_kaffinity(group_processors_mask);
+			const uint32_t processor_global_index =
+				global_proc_index_per_group[group_id] + processor_id_in_group;
+
+			store_cache_info_per_processor(
+				processors, processor_global_index,
+				info, current_cache);
+
+			/* Clear the bits in affinity mask, lower the least set bit. */
+			group_processors_mask &= (group_processors_mask - 1);
+		}
+	}
+	return true;
+}
+
+static void store_package_info_per_processor(
+	struct cpuinfo_processor* processors,
+	const uint32_t processor_global_index,
+	const uint32_t package_id,
+	const uint32_t group_id,
+	const uint32_t processor_id_in_group)
+{
+	processors[processor_global_index].windows_group_id =
+		(uint16_t) group_id;
+	processors[processor_global_index].windows_processor_id =
+		(uint16_t) processor_id_in_group;
+
+	/* As we're counting the number of packages now, we haven't allocated memory for
+	 * cpuinfo_packages yet, so we only set the package pointer's offset now.
+	 */
+	processors[processor_global_index].package =
+		(const struct cpuinfo_package*) NULL + package_id;
+}
+
+void store_core_info_per_processor(
+	struct cpuinfo_processor* processors,
+	const uint32_t processor_global_index,
+	const uint32_t core_id,
+	PSYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX core_info,
+	struct cpuinfo_core* cores,
+	const struct woa_chip_info *chip_info)
+{
+	if (cores) {
+		processors[processor_global_index].core = cores + core_id;
+		cores[core_id].core_id = core_id;
+		get_core_uarch_for_efficiency(
+			chip_info->chip_name, core_info->Processor.EfficiencyClass,
+			&(cores[core_id].uarch), &(cores[core_id].frequency));
+
+		/* We don't have cluster information, so we handle it as
+		 * fixed 1 to (cluster / cores).
+		 * Set the cluster offset ID now, as soon as we have the
+		 * cluster base address, we'll set the absolute address.
+		 */
+		processors[processor_global_index].cluster =
+			(const struct cpuinfo_cluster*) NULL + core_id;
+	}
+}
+
+static void store_cache_info_per_processor(
+	struct cpuinfo_processor* processors,
+	const uint32_t processor_global_index,
+	PSYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX info,
+	struct cpuinfo_cache* current_cache)
+{
+	if (current_cache->processor_start > processor_global_index) {
+		current_cache->processor_start = processor_global_index;
+	}
+	current_cache->processor_count++;
+
+	switch(info->Cache.Level) {
+		case 1:
+			switch (info->Cache.Type) {
+				case CacheInstruction:
+					processors[processor_global_index].cache.l1i = current_cache;
+				break;
+				case CacheData:

```

Comment:
Just a nit: Could you add 'CacheUnified' and 'CacheTrace' to this switch as well?